metadata-io/src/main/java/com/linkedin/metadata/entity/UrnValidator.java

```diff
@@ -0,0 +1,86 @@
+//
+// Source code recreated from a .class file by IntelliJ IDEA
+// (powered by FernFlower decompiler)
+//
+
+package com.linkedin.metadata.entity;
+
+import com.linkedin.common.urn.Urn;
+import com.linkedin.data.message.Message;
+import com.linkedin.data.schema.DataSchema.Type;
+import com.linkedin.data.schema.NamedDataSchema;
+import com.linkedin.data.schema.PathSpec;
+import com.linkedin.data.schema.validator.Validator;
+import com.linkedin.data.schema.validator.ValidatorContext;
+import com.linkedin.data.template.RecordTemplate;
+import com.linkedin.metadata.models.EntitySpec;
+import com.linkedin.metadata.models.RelationshipFieldSpec;
+import com.linkedin.metadata.models.registry.EntityRegistry;
+import com.linkedin.metadata.utils.EntityKeyUtils;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Setter;
+
+
+public class UrnValidator implements Validator {
+
+  private final EntityRegistry _entityRegistry;
+  @Setter
+  private EntitySpec currentEntitySpec = null;
+
+  public UrnValidator(EntityRegistry entityRegistry) {
+    _entityRegistry = entityRegistry;
+  }
+
+  public void validate(ValidatorContext context) {
+    if (currentEntitySpec == null) {
+      throw new IllegalStateException("Current entity spec must be set");
+    }
+    if (Type.TYPEREF.equals(context.dataElement().getSchema().getType()) && ((NamedDataSchema) context.dataElement()
+        .getSchema()).getName().endsWith("Urn")) {
+      try {
+        // Validate Urn matches field type and that it generates a valid key
+        String urnStr = (String) context.dataElement().getValue();
+        Urn urn = Urn.createFromString(urnStr);
+        EntitySpec entitySpec = _entityRegistry.getEntitySpec(urn.getEntityType());
+        RecordTemplate entityKey = EntityKeyUtils.convertUrnToEntityKey(urn,
+            entitySpec.getKeyAspectSpec().getPegasusSchema());
+        NamedDataSchema namedDataSchema = ((NamedDataSchema) context.dataElement().getSchema());
+        Class<? extends Urn> urnClass;
+        try {
+          String schemaName = ((Map<String, String>) namedDataSchema.getProperties().get("java")).get("class");
+          urnClass = (Class<? extends Urn>) Class.forName(schemaName);
+          urnClass.getDeclaredMethod("createFromString", String.class).invoke(null, urnStr);
```

Comment:
Is the purpose of this reflection to support strongly typed urn validation, as opposed to just using Urn.createFromString?

Reply:
Yes! Currently we have explicit typing in the schema (i.e. platform: DatasetUrn), which can pass Urn.createFromString but fails when doing for example schemaMetadata.getPlatform(). This invokes the subtype's createFromString and will fail on yhe insert instead of erroring on the get. If we move to pure Urn in the schema this can be removed and any validations pushed out to however else we want to constrain fields to particular entity types.